src/main/java/de/mediathekview/mserver/crawler/ard/json/ArdMediaArrayToDownloadUrlsConverter.java

```diff
@@ -32,52 +35,153 @@ private ArdMediaArrayToDownloadUrlsConverter() {}
+  }
+  
+  /**
+   * returns the url to use for downloads
+   * uses the following order:
+   * mp4 > m3u8 > Rest
+   * @param aUrls list of possible urls
+   * @return the download url
+   */
+  private static String determineUrl(Set<String> aUrls) {
+    String usedFileType = "";
+    String relevantUrl = "";
+    
+    for (String url: aUrls) {
+      Optional<String> fileType = UrlUtils.getFileType(url);
+      if (fileType.isPresent()) {
+        switch(fileType.get()) {
+          case "mp4":
+            return url;
+          case "m3u8":
+            if (!usedFileType.equals("mp4")) {
+              usedFileType = fileType.get();
+              relevantUrl = url;
+            }
+            break;
+          default:
+            if (usedFileType.isEmpty()) {
+              usedFileType = fileType.get();
+              relevantUrl = url;
+            }
+        }
+      }
+    }
+    
+    return relevantUrl;
+  }
+  
+  private static void parseMediaArray(JsonArray aMediaArray, Map<Resolution, Set<String>> aUrls) {
+    
+    aMediaArray.forEach(mediaEntry -> {
+      final JsonArray mediaStreamArray = mediaEntry.getAsJsonObject()
+              .getAsJsonArray(ELEMENT_MEDIA_STREAM_ARRAY);
+
+      parseMediaStreamArray(mediaStreamArray, aUrls);
+    });    
+  }
+
+  private static void parseMediaStreamArray(JsonArray aMediaStreamArray, Map<Resolution, Set<String>> aUrls) {
+
+    for (int i = 0; i < aMediaStreamArray.size(); i++) {
+      final JsonElement videoElement = aMediaStreamArray.get(i);
+      final String qualityAsText =
+          videoElement.getAsJsonObject().get(ELEMENT_QUALITY).getAsString();
+      
+      Optional<Resolution> quality = getQuality(qualityAsText);
+      if (quality.isPresent()) {
+        String downloadUrl = "";
+        
+        if (videoElement.getAsJsonObject().has(ELEMENT_SERVER)) {
+          final String baseUrl = videoElement.getAsJsonObject().get(ELEMENT_SERVER).getAsString();
+          downloadUrl = videoElementToUrl(videoElement, baseUrl);
+          addUrl(videoElement, downloadUrl, quality.get(), aUrls);
+        } 
+        
+        if (downloadUrl.isEmpty() && videoElement.getAsJsonObject().has(ELEMENT_STREAM)) {
+          JsonElement streamObject = videoElement.getAsJsonObject().get(ELEMENT_STREAM);
+          if (streamObject.isJsonPrimitive()) {
+            final String baseUrl = streamObject.getAsString();
+            downloadUrl= videoElementToUrl(videoElement, baseUrl);
+            addUrl(videoElement, downloadUrl, quality.get(), aUrls);          
+          } else if(streamObject.isJsonArray()) {
+            JsonArray streamArray = streamObject.getAsJsonArray();
+            streamArray.forEach(stream -> {
+              final String baseUrl = stream.getAsString();
+              addUrl(videoElement, baseUrl, quality.get(), aUrls);          
+            });
+          }
+        }
+      }
+    }
+  }
+  
+  private static void addUrl(final JsonElement aVideoElement, 
```

Comment:
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') Remove this unused method parameter "aVideoElement". [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1172)